### PR TITLE
Improve unparsing for `ORDER BY`, `UNION`, Windows functions with Aggregation

### PR DESCRIPTION
**Pull request by sgrebnov**
## Which issue does this PR close?

Improves unparsing to produce valid sql for TPC-DS benchmark queries:

#### Improve UNION unparsing

Fix UNION unparsing for queries where UNION is a subquery

Original query
```sql
select
	channel,
	col_name
from
	(
	select
		'store' as channel,
		'ss_customer_sk' as col_name
	from
		store_sales
union all
	select
		'web' as channel,
		'ws_ship_hdemo_sk' as col_name
	from
		web_sales
    ) as foo
group by
	channel,
	col_name
order by
	channel,
	col_name
```

Before (invalid):
```sql
select
	'store' as "channel",
	'ss_customer_sk' as "col_name"
from
	"store_sales"
union all
select
	'web' as "channel",
	'ws_ship_hdemo_sk' as "col_name"
from
	"web_sales"
order by
	"foo"."channel" asc nulls last,
	"foo"."col_name" asc nulls last
```

After (correct)
```sql
select
	"foo"."channel",
	"foo"."col_name"
from
	(
	select
		'store' as "channel",
		'ss_customer_sk' as "col_name"
	from
		"store_sales"
union all
	select
		'web' as "channel",
		'ws_ship_hdemo_sk' as "col_name"
	from
		"web_sales") as "foo"
group by
	"foo"."channel",
	"foo"."col_name"
order by
	"foo"."channel" asc nulls last,
	"foo"."col_name" asc nulls last
```

#### Fix unparsing for ORDER BY with Aggregation functions 

Improve unparsing to handle cases when `ORDER BY` clause contains  aggregation functions that needs to be unprojected, otherwise it will result in unparsed/incorrect sql  - `Referenced column "sum(catalog_returns.cr_net_loss)" not found in FROM clause` / `Referenced column "count(DISTINCT cs1.cs_order_number)" not found in FROM clause`

For example TPC-DS Q42
>Referenced column "sum(store_sales.ss_ext_sales_price)" not found in FROM clause
```sql
select  dt.d_year
 	,item.i_category_id
 	,item.i_category
 	,sum(ss_ext_sales_price)
 from 	date_dim dt
 	,store_sales
 	,item
 where dt.d_date_sk = store_sales.ss_sold_date_sk
 	and store_sales.ss_item_sk = item.i_item_sk
 	and item.i_manager_id = 1
 	and dt.d_moy=11
 	and dt.d_year=1998
 group by 	dt.d_year
 		,item.i_category_id
 		,item.i_category
 order by       sum(ss_ext_sales_price) desc,dt.d_year
 		,item.i_category_id
 		,item.i_category
 LIMIT 100 ;
```

#### Improve scalar functions in ORDER BY unparsing

Improve unparsing to correctly handle scalar functions in ORDER BY. If SELECT and ORDER BY contain the same expression with a scalar function, the ORDER BY expression will be replaced by a Column expression (e.g., "substr(customer.c_last_name, Int64(0), Int64(5))"), and we need to transform it back to the actual expression. This is done for scalar function expressions only.

```sql
  select c_last_name,c_first_name, substr(c_last_name,0,5)
  from customer
  order by substr(c_last_name,0,5)
  limit 10;
```

Before
```sql
select
	"customer"."c_last_name",
	"customer"."c_first_name",
	substr("customer"."c_last_name",0,5)
from
	"customer"
order by
	"substr(customer.c_last_name,Int64(0),Int64(5))" asc nulls last
limit 10
```

After
```sql
select
	"customer"."c_last_name",
	"customer"."c_first_name",
	substr("customer"."c_last_name",0,5)
from
	"customer"
order by
	substr("customer"."c_last_name",0,5) asc nulls last
limit 10
```

#### Fix unparsing for complex Window functions with Aggregation

Improve `unproject_agg_exprs`  to unparse complex Window functions with aggregation when aggregation functions present in `When/Then and Order By parts`.

For example `case when grouping(i_class) = 0 then i_category end` and  `order by sum(ws_net_paid) desc` in Window `rank()`:

```sql
select
    sum(ws_net_paid) as total_sum
   ,i_category
   ,i_class
   ,grouping(i_category)+grouping(i_class) as lochierarchy
   ,rank() over (
 	partition by grouping(i_category)+grouping(i_class),
 	case when grouping(i_class) = 0 then i_category end
 	order by sum(ws_net_paid) desc) as rank_within_parent
 from
    web_sales
   ,date_dim       d1
   ,item
 where
    d1.d_month_seq between 1205 and 1205+11
 and d1.d_date_sk = ws_sold_date_sk
 and i_item_sk  = ws_item_sk
 group by rollup(i_category,i_class)
 order by
   lochierarchy desc,
   case when lochierarchy = 0 then i_category end,
   rank_within_parent
  LIMIT 100;
```

## Are these changes tested?

Added automated tests

## Are there any user-facing changes?

Unparser will generate valid sql for complex Window functions  and Order By w/ aggregations and Union and Order By with Aggregation functions

### CODE REVIEW

File: datafusion/sql/src/unparser/expr.rs [hunk 1]
```diff
@@ -76,11 +76,6 @@ pub fn expr_to_sql(expr: &Expr) -> Result<ast::Expr> {
     unparser.expr_to_sql(expr)
 }
 
-pub fn sort_to_sql(sort: &Sort) -> Result<ast::OrderByExpr> {
```
Comment (sgrebnov) [inline datafusion/sql/src/unparser/expr.rs:79]: Removed in favor of unparser sort_to_sql which respects `supports_nulls_first_in_sort` dialect setting

File: datafusion/sql/src/unparser/utils.rs [hunk 2]
```diff
@@ -118,21 +119,11 @@ pub(crate) fn unproject_agg_exprs(
...
-                                if let Some(expr) = find_agg_expr(agg, c)? {
-                                    *arg = expr.clone();
-                                }
-                            }
-                            Ok::<(), DataFusionError>(())
-                        })?;
-                    }
-                    Ok(Transformed::yes(unprojected_expr))
+                    // Window function can contain an aggregation columns, e.g., 'avg(sum(ss_sales_price)) over ...' that needs to be unprojected
+                    return Ok(Transformed::yes(unproject_agg_exprs(&unprojected_expr, agg, None)?));
```
Comment (sgrebnov) [inline datafusion/sql/src/unparser/utils.rs:126]: This unprojects complex Window functions including `order_by`, etc not just `args`.

```
ub struct WindowFunction {
    /// Name of the function
    pub fun: WindowFunctionDefinition,
    /// List of expressions to feed to the functions as arguments
    pub args: Vec<Expr>,
    /// List of partition by expressions
    pub partition_by: Vec<Expr>,
    /// List of order by expressions
    pub order_by: Vec<Sort>,
    /// Window frame
    pub window_frame: WindowFrame,
    /// Specifies how NULL value is treated: ignore or respect
    pub null_treatment: Option<NullTreatment>,
}
```

File: datafusion/sql/src/unparser/plan.rs [hunk 3]
```diff
@@ -542,6 +553,11 @@ impl Unparser<'_> {
                     );
                 }
 
+                // Covers cases where the UNION is a subquery and the projection is at the top level
+                if select.already_projected() {
```
Comment (sgrebnov) [inline datafusion/sql/src/unparser/plan.rs:557]: Small improvement to correctly handle UNION unparsing for queries where UNION is a subquery

File: datafusion/sql/src/unparser/utils.rs [hunk 4]
```diff
@@ -200,6 +191,54 @@ fn find_window_expr<'a>(
         .find(|expr| expr.schema_name().to_string() == column_name)
 }
 
+/// Transforms a Column expression into the actual expression from aggregation or projection if found.
+/// This is required because if an ORDER BY expression is present in an Aggregate or Select, it is replaced
+/// with a Column expression (e.g., "sum(catalog_returns.cr_net_loss)"). We need to transform it back to
+/// the actual expression, such as sum("catalog_returns"."cr_net_loss").
+pub(crate) fn unproject_sort_expr(
```
Comment (sgrebnov) [inline datafusion/sql/src/unparser/utils.rs:198]: Might be better to update internal implementation to rely on `.transform`

File: datafusion/sql/tests/cases/plan_to_sql.rs [hunk 5]
```diff
@@ -893,3 +908,49 @@ fn test_with_offset0() {
...
+        r#"SELECT id, first_name,
+        SUM(id) AS total_sum,
+        SUM(id) OVER (PARTITION BY first_name ROWS BETWEEN 5 PRECEDING AND 2 FOLLOWING) AS moving_sum,
+        MAX(SUM(id)) OVER (PARTITION BY first_name ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING) AS max_total,
+        rank() OVER (PARTITION BY grouping(id) + grouping(age), CASE WHEN grouping(age) = 0 THEN id END ORDER BY sum(id) DESC) AS rank_within_parent_1,
+        rank() OVER (PARTITION BY grouping(age) + grouping(id), CASE WHEN (CAST(grouping(age) AS BIGINT) = 0) THEN id END ORDER BY sum(id) DESC) AS rank_within_parent_2
+        FROM person
+        GROUP BY id, first_name;"#,
+        r#"SELECT person.id, person.first_name,
+sum(person.id) AS total_sum, sum(person.id) OVER (PARTITION BY person.first_name ROWS BETWEEN '5' PRECEDING AND '2' FOLLOWING) AS moving_sum,
```
Comment (goldmedal) [inline datafusion/sql/tests/cases/plan_to_sql.rs:949]: I noticed the original sql in L942 is
```sql
SUM(id) OVER (PARTITION BY first_name ROWS BETWEEN 5 PRECEDING AND 2 FOLLOWING) AS moving_sum
```
However, the long literals `5` and `2` are converted to string literals `'5'` and `'2'`. I think the results are wrong but I'm not sure if it's an issue caused by this PR. If not, I think we can file an issue and fix it in another PR 🤔

**Review by goldmedal — COMMENTED** [review]
Thanks @sgrebnov,  I think overall is good but one question and minor suggestions.

File: datafusion/sql/tests/cases/plan_to_sql.rs (hunk 5 above)
Comment (sgrebnov) [reply datafusion/sql/tests/cases/plan_to_sql.rs:949]: @goldmedal  - thank you very much for review - I'll double check this plus simplify the code as suggested above later today and will reply back!

Comment (sgrebnov) [reply datafusion/sql/tests/cases/plan_to_sql.rs:949]: @goldmedal - updated code + checked repo master behavior to confirm  literals 5 and 2 are converted to string literals '5' and '2' was NOT introduced by this change. Same behavior so this is safe to merge and I'll investigate 

https://github.com/apache/datafusion/issues/12982

Comment (goldmedal) [reply datafusion/sql/tests/cases/plan_to_sql.rs:949]: Thanks @sgrebnov !

**Review by goldmedal — APPROVED** [review]
👍

**Review by alamb — APPROVED** [review]
Thank you very much @sgrebnov and very much appreciate the help reviewing @goldmedal

**Pull request merged**